Better handling of UTF-8 paths

viettaml opened the pull request:
Remove calls to qPrintable which destroys non-ANSI path names on Windows. Replace with QString::toUtf8() and modify low-level file handling code to use _wfopen() and other wide-char functions on Windows to support non-ASCII paths. (I did an experiment to convince myself that fopen() doesn't do UTF-8, and that _wfopen() correctly creates a file with a non-ASCII filename starting from a UTF-8 name.) Add ufopen() function as a UTF-8 wrapper for fopen(). Change inifile_init() filename argument type to QString.

Leave serial port paths alone ("if your serial port path has non-ANSI characters, you're going to have a bad time").

viettaml committed:
Better handling of UTF-8 paths

Remove gpsbabel calls to qPrintable which destroys non-ANSI path names on Windows. Replace with QString::toUtf8() and modify low-level file handling code to use _wfopen() and other wide-char functions on Windows to support non-ASCII paths. (I did an experiment to convince myself that fopen() doesn't do UTF-8, and that _wfopen() correctly creates a file with a non-ASCII filename starting from a UTF-8 name.) Add ufopen() function as a UTF-8 wrapper for fopen(). Change inifile_init() filename argument type to QString.

Leave serial port paths alone ("if your serial port path has non-ANSI characters, you're going to have a bad time").

viettaml committed:
Fix valgrind mismatched free/malloc error.

tsteven4 commented on gbfile.cc:
```diff
@@ -538,8 +538,7 @@ gbfopen(const QString filename, const char* mode, const char* module)
     file->filewrite = memapi_write;
   } else {
-    /* Be careful to convert back to local8Bit for these c based APIS */
-    file->name = xstrdup(qPrintable(filename));
```
If you merge up travis-ci should show a problem when it runs test_encoding due to not converting back to local8bit.

viettaml committed:
Correctly encode output paths in native locale on non-Windows

On Mac/Linux, convert output file path from internal UTF-8 to local encoding (which may be UTF-8 or something else) when creating the file. Fixes test_encoding failure.

viettaml committed:
Support Unicode paths for .gz files on Windows

Use zlib's Windows-specific gzopen_w() to support Unicode paths.

tsteven4 commented on gbfile.cc:
```diff
@@ -84,7 +85,13 @@ gzapi_open(gbfile* self, const char* mode)
+    // On Windows, use gzopen_w() to handle Unicode paths.
+    std::wstring wname = QString(self->name).toStdWString();
+    self->handle.gz = gzopen_w(wname.c_str(), openmode);
+#else
```
if you remove the qPrintable at in gbfile.cc at old line 542 you will need to add one at new line 93 like you did in util.cc at new line 274.

viettaml replied:
Thanks, fixed (I think).

viettaml replied:
Also worth noting that with Qt 5.7 and later we could use qUtf16Printable() to convert straight from a QString() to a temporary const wchar_t*.

viettaml committed:
Convert UTF-8 to local encoding when opening gzip file on non-Windows

This should fix the test_encoding failure on Travis-CI.

Also, change Windows side to manually allocating wchar_t array; although a bit less elegant, it's very localized and avoids introducing a whole new dependency on std::wstring.

viettaml committed:
Use explicit char* to QString conversion in qPrintable

The implicit conversion works, but I prefer the an explicit one (if only as a reminder that we should convert all internal strings to QString).

viettaml committed:
Add comment to QString-to-wchar_t[] conversion

... since it may not be immediately obvious what that line of code does.

tsteven4 commented on shape.cc:
```diff
@@ -56,12 +56,12 @@ arglist_t shp_args[] = {
 my_rd_init(const QString& fname)
 {
-  ihandle = SHPOpen(qPrintable(fname), "rb");
+  ihandle = SHPOpen(fname.toUtf8(), "rb");
```
This will break support for non-utf8 unix, and it will not fix windows in cases where windows fopen would not work.  Under normal circumstances these calls lead to fopen.  To fix windows it seems necessary to call SASetupUtf8Hooks at least after SHPopen, DBFOpen, SHPcreate, DBFCreate, DBFCreateEx, DBFCloneEmpty (any where that leads to SASetupDefaultHooks being called in shapelib) to get windows utf8 support.  SASetupUtf8Hooks can only be called if SHPAPI_UTF8_HOOKS is defined, and it will only work if SHPAPI_WINDOWS is also defined.
We also have *NO* test cases.  However it should be relatively easy to see if filenames like test_encoding_fileαβγ.shx and test_encoding_fileαβγ.shp at least find files on input and write them on output.
This comment applies to all changes in this file.

tsteven4 commented on util.cc:
```diff
@@ -258,6 +258,23 @@ xfopen(const char* fname, const char* type, const char* errtxt)
+  return fopen(qPrintable(QString(fname)), mode);
+#endif
+}
+
```
I think it would be better to use one method for QString -> wchar_t conversion.  gbfile is now using a different method that util.cc.  Let's settle on one.

viettaml replied:
Done.
kml.cc and util.cc are using the QString::utf16() to const wchar_t* cast, so I've updated gbfile.cc to use that too.

viettaml committed:
Merge branch 'master' into viettaml-patch-5

viettaml committed:
Merge branch 'master' into viettaml-patch-5

viettaml committed:
Fix merge conflict resolution error

viettaml committed:
Fix merge conflict (again)

viettaml committed:
Revert shape.cc

viettaml committed:
Update gbfile.cc

viettaml committed:
Merge branch 'viettaml-patch-5' of github.com:viettaml/gpsbabel into viettaml-patch-5

tsteven4 commented on mtk_logger.cc:
```diff
@@ -617,7 +617,7 @@ static void mtk_read()
     dsize = 0;
     init_scan = 0;
     rename(TEMP_DATA_BIN, TEMP_DATA_BIN_OLD);
-    dout = fopen(TEMP_DATA_BIN, "wb");
```
should we use QFile::rename here like we do in other formats?  I suspect rename may not work on WIN32 with utf8 names.

viettaml committed:
Update mkt_logger.cc to use QString tempfile paths.

viettaml committed:
Change ufopen() to take QString filename argument.
Where converting from char*, use an explicit QString::fromUtf8() conversion rather than the implicit QString(const char*) constructor.

tsteven4 commented on garmin_device_xml.cc:
```diff
@@ -127,7 +127,7 @@ const gdx_info*
 {
   // Test file open-able before gb_open gets a chance to fatal().
-  FILE* fin = fopen(fname, "r");
+  FILE* fin = ufopen(QString::fromUtf8(fname), "r");
```
The comment at line 129 was before xmlgenric was converted to Qt.  Now, instead of having gb_open fatal I think xml_read would fatal in gpsbabel::File::open without a check to make sure the file exists.  When xmlgeneric was converted to Qt we picked up an implicit conversion to QString in xml_init using fromUtf8.  So we already had assumed fname was utf8, but in reality it isn't.  One path goes to get_garmin_mountpoints in jeeps/gpsusbwin.cc which calls GetLogicalDriveStringsA, the ANSI version.  So your PR is no worse than the HEAD, but they probably both fail with certain path names.  I am ready to let that go under the justification that you have done no harm and some good.  This bug has existed for a while without the world collapsing and I would like to move this PR along.

But instead of ufopen I would suggest using QFile::exists(fname).  I think the fromUtf8 conversion will be done implicitly if you don't spell it out which would work out better if gdx_read every takes a QString.  To me the explicit conversion makes me think you know your char string is utf8 and you don't want people to wonder if the implicit conversion is correct, or your getting the code ready for QT_NO_CAST_FROM_ASCII (which would probably be a very long row for us to hoe).  Your thoughts?

Please change the gb_open comment to xml_read.  Please also comment about the potential failure mode described above with a "TODO:"

viettaml replied:
Thanks (I blindly converted the fopen calls without paying particular attention to what they were used for), will fix.

tsteven4 commented on inifile.cc:
```diff
@@ -91,11 +91,12 @@ open_gpsbabel_inifile()
   char* envstr;
   gbfile* res = NULL;
 
+  // TODO(viettaml): use _wgetenv() on Windows, convert to QString.
```
you would hope qgetenv would handle this for us, but the doc explictly warns that it won't work.

viettaml replied:
Yeah, I looked into this as well and it looks like it might be fixed in the very latest Qt 5.10, but that doesn't help us (or me, as Earth uses Qt 5.4): https://bugreports.qt.io/browse/QTBUG-41006

The best thing to do would be to implement a wrapper, say xgetenv() in defs.h/utils.cc, doing the right conversion for each OS and returning a QString.

Your thoughts on whether or not that should go in a separate follow-up PR?

tsteven4 replied:
If you will please do that in a separate PR.

viettaml committed:
Use QFile::exists() and QFile::open() rather than ufopen() to verify a file exists and can be opened.

viettaml commented on garmin_device_xml.cc:
```diff
@@ -127,10 +128,7 @@ const gdx_info*
-
-  if (fin) {
-    fclose(fin);
+  if (QFile::exists(fname) && QFile(fname).open(QIODevice::ReadOnly)) {
```
I realized QFile::exists() by itself isn't strictly enough, since just because a file exists (and can be seen) doesn't mean it can be opened. So I added a QFile::open() call as well. Come to think of it, I suppose that makes the QFile::exists() call redundant, assuming QFile::open() will fail "softly" without generating an exception/abort/nasty output.

viettaml committed:
Remove redundant call to QFile::exists(). Trivial include cleanup in main.cc.

viettaml committed:
Remove superfluous QVector include.

viettaml committed:
Merge branch 'master' of github.com:viettaml/gpsbabel into viettaml-patch-5

GPSBabelDeveloper commented on kml.cc:
```diff
@@ -541,7 +541,8 @@ kml_wr_deinit()
   if (!posnfilenametmp.isEmpty()) {
 #if __WIN32__
-    MoveFileExA(qPrintable(posnfilenametmp), qPrintable(posnfilename),
+    MoveFileExW((const wchar_t*) posnfilenametmp.utf16(),
```
Tam, I think this has been through some mechanical changes without testing. This code is to do an atomic rename so that the freshly written KML (that ends in </kml> is always readable.  Is this MoveFileExW() needed at all? That QFile::rename below should do the trick everywhere, right?

This chunk predates having QtCore.

tsteven4 replied:
Ugg.  I agree with:
> I think this has been through some mechanical changes without testing.

There are bugs here.  
1. I introduced one with 15f8b8b6.  
See https://bugreports.qt.io/browse/QTBUG-2082
 >                   // QFile.rename will always fail if the target
 >                   // file exists. Posix rename wouldn't
2. It seems to me that on windows when we use MoveFileEx* there is no need to do the QFile::rename, and it is guaranteed to fail as posnfilename will exist due to MoveFileEx*.

The bug report makes it sound like QSaveFile does what we want.  The QSaveFile doc says
>QSaveFile creates a temporary file in the same directory as the final file and atomically renames it. However this is not possible if the directory permissions do not allow creating new files. In order to preserve atomicity guarantees, open() fails when it cannot create the temporary file.

tsteven4 replied:
I think QSaveFile would cause some ripples that we may not want to hold the PR for.  How about
#if __WIN32__
    MoveFileExW((const wchar_t*) posnfilenametmp.utf16(),
                (const wchar_t*) posnfilename.utf16(),
                MOVEFILE_REPLACE_EXISTING);
#else
    rename(qPrintable(posnfilenametmp), qPrintable(posnfilename));
#endif

GPSBabelDeveloper reviewed (commented):
I've lost the beat with the thread, but this changeset seems to solve a number of pretty esoteric and hard to diagnose issues in this area. The two of you have worked well together to come to good resolutions and changes. Thank you.

I'm OK with this change going in, but I'm going to defer the final approval to tsteven4. I don't think I have any pending merge conflicts in this area, so I'd welcome it landing in the next few days.

Thanx for being so meticulous in working through this stuff.  These are definitely areas where square pegs were forcibly hammered into round holes. :-)

tsteven4 committed:
Merge branch 'master' into viettaml-patch-5

tsteven4 committed:
Merge branch 'master' into viettaml-patch-5

tsteven4 committed:
add newline at EOF.

tsteven4 commented on mtk_logger.cc:
```diff
@@ -612,14 +609,15 @@ static void mtk_read()
-    rename(TEMP_DATA_BIN, TEMP_DATA_BIN_OLD);
-    dout = fopen(TEMP_DATA_BIN, "wb");
+    QFile::rename(TEMP_DATA_BIN, TEMP_DATA_BIN_OLD);
+    dout = ufopen(TEMP_DATA_BIN, "wb");
```
QFile::rename will fail if TEMP_DATA_BIN_OLD exists without overwriting it.  The behavior of rename may be implementation dependent.  See https://wiki.sei.cmu.edu/confluence/display/c/FIO10-C.+Take+care+when+using+the+rename%28%29+function
